fpga_interchange/site_routing_cache.h

```diff
@@ -0,0 +1,133 @@
+ *  WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
+ *  ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+ *  OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+ *
+ */
+
+#ifndef SITE_ROUTING_CACHE_H
+#define SITE_ROUTING_CACHE_H
+
+#include "PhysicalNetlist.capnp.h"
+#include "nextpnr_namespaces.h"
+#include "site_arch.h"
+#include "site_routing_storage.h"
+
+NEXTPNR_NAMESPACE_BEGIN
+
+struct SiteRoutingSolution
+{
+    void store_solution(const SiteArch *ctx, const RouteNodeStorage *node_storage, const SiteWire &driver,
+                        std::vector<size_t> solutions);
+    void verify(const SiteArch *ctx, const SiteNetInfo &net);
+
+    void clear()
+    {
+        solution_offsets.clear();
+        solution_storage.clear();
+        solution_sinks.clear();
+    }
+
+    size_t num_solutions() const { return solution_sinks.size(); }
+
+    const SiteWire &solution_sink(size_t solution) const { return solution_sinks.at(solution); }
+
+    std::vector<SitePip>::const_iterator solution_begin(size_t solution) const
+    {
+        NPNR_ASSERT(solution + 1 < solution_offsets.size());
+        return solution_storage.begin() + solution_offsets.at(solution);
+    }
+
+    std::vector<SitePip>::const_iterator solution_end(size_t solution) const
+    {
+        NPNR_ASSERT(solution + 1 < solution_offsets.size());
+        return solution_storage.begin() + solution_offsets.at(solution + 1);
+    }
+
+    std::vector<size_t> solution_offsets;
+    std::vector<SitePip> solution_storage;
+    std::vector<SiteWire> solution_sinks;
+};
+
+struct SiteRoutingKey
+{
+    int32_t tile_type;
+    int32_t site;
+    // The net type matters for site routing. Legal routes for VCC/GND/SIGNAL
+    // nets are different.
+    PhysicalNetlist::PhysNetlist::NetType net_type;
+    SiteWire::Type driver_type;
+    int32_t driver_index;
+    std::vector<SiteWire::Type> user_types;
+    std::vector<int32_t> user_indicies;
+
+    bool operator==(const SiteRoutingKey &other) const
+    {
+        return tile_type == other.tile_type && site == other.site && net_type == other.net_type &&
+               driver_type == other.driver_type && driver_index == other.driver_index &&
+               user_types == other.user_types && user_indicies == other.user_indicies;
+    }
+    bool operator!=(const SiteRoutingKey &other) const
+    {
+        return tile_type != other.tile_type || site != other.site || net_type != other.net_type ||
+               driver_type != other.driver_type || driver_index != other.driver_index ||
+               user_types != other.user_types || user_indicies != other.user_indicies;
+    }
+
+    static SiteRoutingKey make(const SiteArch *ctx, const SiteNetInfo &site_net);
+};
+
+NEXTPNR_NAMESPACE_END
+
+template <> struct std::hash<NEXTPNR_NAMESPACE_PREFIX SiteRoutingKey>
+{
+    std::size_t operator()(const NEXTPNR_NAMESPACE_PREFIX SiteRoutingKey &key) const noexcept
```

Comment:
could this use a precomputed hash for better performance like `TypeWireSet` ?

Reply:
Believe or not, once I implemented the site routing cache, this was no longer in the hot path.  Remaining `isBelLocationValid` hot path is (values from murax run):
 - is_cell_valid_constraints: 11.84s
 - remap_luts: 87s
 - route_site: 33.18s
 - SiteArch: 16.56s

Once I fix the remaining issues preventing the design from getting to a DCP, I have good ideas on how to improve most of those numbers.